c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+    let (before, rest) = rts.split_at(i);
+
+    // Collect nodes from the front of `rts` until we find one that comes strictly after `span`.
+    let j = rest
+        .iter()
+        .position(|rt| rt.span.lo() >= hi)
+        .unwrap_or(rest.len());
+    let (overlap, after) = rest.split_at(j);
+
+    (before, overlap, after)
+}
+
+struct Emitter<'a, F> {
+    file: &'a SourceFile,
+    emit: &'a mut F,
+}
+
+impl<'a, F: FnMut(&str)> Emitter<'a, F> {
+    fn emit_str(&mut self, s: &str) {
+        (self.emit)(s);
+    }
+
+    fn emit_parenthesized(&mut self, cond: bool, f: impl FnOnce(&mut Self)) {
+        if cond {
+            self.emit_str("(");
+        }
+        f(self);
+        if cond {
+            self.emit_str(")");
+        }
+    }
+
+    fn emit_rewrite(
+        &mut self,
+        rw: &Rewrite,
+        prec: usize,
+        emit_expr: &mut impl FnMut(&mut Self),
+        emit_subexpr: &mut impl FnMut(&mut Self, Span),
+    ) {
+        match *rw {
+            Rewrite::Identity => self.emit_parenthesized(true, |slf| {
+                emit_expr(slf);
+            }),
+            Rewrite::Sub(_, span) => self.emit_parenthesized(true, |slf| {
+                emit_subexpr(slf, span);
+            }),
+
+            Rewrite::Ref(ref rw, mutbl) => self.emit_parenthesized(prec > 2, |slf| {
+                match mutbl {
+                    Mutability::Not => slf.emit_str("&"),
+                    Mutability::Mut => slf.emit_str("&mut "),
+                }
+                slf.emit_rewrite(rw, 2, emit_expr, emit_subexpr);
+            }),
+            Rewrite::AddrOf(ref rw, mutbl) => {
+                match mutbl {
+                    Mutability::Not => self.emit_str("core::ptr::addr_of!"),
+                    Mutability::Mut => self.emit_str("core::ptr::addr_of_mut!"),
+                }
+                self.emit_parenthesized(true, |slf| {
+                    slf.emit_rewrite(rw, 0, emit_expr, emit_subexpr);
+                });
+            }
+            Rewrite::Deref(ref rw) => self.emit_parenthesized(prec > 2, |slf| {
+                slf.emit_str("*");
+                slf.emit_rewrite(rw, 2, emit_expr, emit_subexpr);
+            }),
+            Rewrite::Index(ref arr, ref idx) => self.emit_parenthesized(prec > 3, |slf| {
+                slf.emit_rewrite(arr, 3, emit_expr, emit_subexpr);
+                slf.emit_str("[");
+                slf.emit_rewrite(idx, 0, emit_expr, emit_subexpr);
+                slf.emit_str("]");
+            }),
+            Rewrite::SliceTail(ref arr, ref idx) => self.emit_parenthesized(prec > 3, |slf| {
+                slf.emit_rewrite(arr, 3, emit_expr, emit_subexpr);
+                slf.emit_str("[");
+                // Rather than figure out the right precedence for `..`, just force
+                // parenthesization in this position.
+                slf.emit_rewrite(idx, 999, emit_expr, emit_subexpr);
+                slf.emit_str(" ..]");
```

Comment:
is it valid (or even possible) to emit parenthesized expressions within things like this?

Reply:
In a `..` range?  I'm pretty sure.

Reply:
Yes, it's legal to do things like `(1 + 2) .. (3 + 4)`